modules/oe_editorial_corporate_workflow/src/EntityStateTransitionBatch.php

```diff
@@ -0,0 +1,176 @@
+   *   The entity type manager.
+   * @param \Drupal\content_moderation\ModerationInformationInterface $moderationInformation
+   *   The moderation information service.
+   * @param \Drupal\Component\Datetime\TimeInterface $time
+   *   The time service.
+   * @param \Drupal\Core\Session\AccountInterface $currentUser
+   *   The current user.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   The messenger service.
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager, ModerationInformationInterface $moderationInformation, TimeInterface $time, AccountInterface $currentUser, MessengerInterface $messenger) {
+    $this->entityTypeManager = $entityTypeManager;
+    $this->moderationInformation = $moderationInformation;
+    $this->time = $time;
+    $this->currentUser = $currentUser;
+    $this->messenger = $messenger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity_type.manager'),
+      $container->get('content_moderation.moderation_information'),
+      $container->get('datetime.time'),
+      $container->get('current_user'),
+      $container->get('messenger')
+    );
+  }
+
+  /**
+   * Executes a single transition to a state, creating a new entity revision.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The entity being transitioned.
+   * @param array $transitions
+   *   The original list of transitions to execute.
+   * @param string $revision_log_message
+   *   The revision log message.
+   * @param array $context
+   *   The batch context.
+   */
+  public function execute(ContentEntityInterface $entity, array $transitions, string $revision_log_message, array &$context): void {
+    // Initialise the sandbox if needed.
+    if (!isset($context['sandbox']['current_revision'])) {
+      $context['sandbox']['current_revision'] = $entity;
+      $context['sandbox']['transitions'] = $transitions;
+      $context['sandbox']['total'] = count($transitions);
+    }
+
+    $entity = $context['sandbox']['current_revision'];
+    $to_state = array_shift($context['sandbox']['transitions']);
+
+    // Create a new revision for the transition change and save the entity.
+    /** @var \Drupal\Core\Entity\ContentEntityStorageInterface $storage */
+    $storage = $this->entityTypeManager->getStorage($entity->getEntityTypeId());
+    $entity = $storage->createRevision($entity, $entity->isDefaultRevision());
+    // Set the next state id.
+    $entity->set('moderation_state', $to_state);
+
+    if ($entity instanceof RevisionLogInterface) {
+      $entity->setRevisionCreationTime($this->time->getRequestTime());
+      $entity->setRevisionLogMessage($revision_log_message);
+      $entity->setRevisionUserId($this->currentUser->id());
+    }
+    $entity->save();
+
+    $context['sandbox']['current_revision'] = $entity;
```

Comment:
Is it a good idea to serialize the entire entity? It occupies lots of space, needlessly in the batch table. Maybe we can just store the revision ID?

Reply:
Thing is batch operations are run one after each other in the same request if possible, if the sum of the time spent is not above 1 second. If we don't serialize the entity, we need to load it at each op, which might be longer.
I also serialised the current entity status to keep all the possible extra "temporary" properties that might be assigned to it, that would be lost if we reload the entity.